gstd/src/common/reservations.rs

```diff
@@ -0,0 +1,215 @@
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with this program. If not, see <https://www.gnu.org/licenses/>.
+
+use super::primitives::ReservationId;
+use crate::{errors::Result, exec, prelude::*};
+
+struct Reservation {
+    id: ReservationId,
+    amount: u64,
+    valid_until: u32,
+}
+
+/// Result of [`Reservations::try_take_reservation`](Reservations::try_take_reservation)
+pub enum TakeReservationResult {
+    /// Appropriate reservation found.
+    Sufficient {
+        /// [`ReservationId`](ReservationId) correspondinfg to the found
+        /// reservation.
+        id: ReservationId,
+        /// amount of gas stored in found reservation
+        amount: u64,
+    },
+    /// Appropriate reservation hasn't been found so the biggest available
+    /// reserved amount is returned.
+    Insufficient(u64),
+    /// There's no reservations inside [`Reservations`](Reservations)
+    NoReservations,
+}
+
+/// Reservation manager.
+///
+/// The manager is used to control multiple gas reservations
+/// across executions.
+///
+/// # Examples
+///
+/// Create gas reservations inside `init` and use them inside `handle`
+///
+/// ```
+/// use gstd::{msg, prelude::*, Reservations, TakeReservationResult};
+///
+/// static mut RESERVATIONS: Option<Reservations> = None;
+///
+/// extern "C" fn init() {
+///     unsafe {
+///         RESERVATIONS = Some(Reservations::new());
+///
+///         RESERVATIONS
+///             .as_mut()
+///             .unwrap()
+///             .reserve(200_000, 50)
+///             .expect("failed to reserve gas");
+///         RESERVATIONS
+///             .as_mut()
+///             .unwrap()
+///             .reserve(100_000, 100)
+///             .expect("failed to reserve gas");
+///         RESERVATIONS
+///             .as_mut()
+///             .unwrap()
+///             .reserve(50_000, 30)
+///             .expect("failed to reserve gas");
+///     }
+/// }
+///
+/// extern "C" fn handle() {
+///     let reservation = unsafe { RESERVATIONS.as_mut().unwrap().try_take_reservation(100_000) };
+///     match reservation {
+///         TakeReservationResult::Sufficient { id, amount } => {
+///             msg::send_bytes_from_reservation(
+///                 id,
+///                 msg::source(),
+///                 b"send_bytes_from_reservation",
+///                 0,
+///             )
+///             .expect("Failed to send message from reservation");
+///         }
+///         TakeReservationResult::Insufficient(_) | TakeReservationResult::NoReservations => {
+///             msg::send_bytes(msg::source(), b"send_bytes", 0).expect("Failed to send message");
+///         }
+///     }
+/// }
+/// ```
+///
+/// # See also
+/// - [`ReservationId`](ReservationId) is used to reserve and unreserve gas
+/// for program execution later.
+#[derive(Default)]
+pub struct Reservations {
+    reservations: Vec<Reservation>,
+}
```

Comment:
Seems more neat for me
```suggestion
pub struct Reservations(Vec<Reservation>);
```